fDefault reconnect strategy uses exponential backoff and jitter

qsymmachus opened the pull request:
### Description

The current default retry strategy increases delay linearly with no jitter. This can lead to a thundering herd problem if many clients lose their connection at once, which is common during events like a Redis upgrade.

A retry strategy that uses exponential backoff plus a random jitter helps mitigate this risk and [is recommended](https://aws.amazon.com/blogs/database/best-practices-redis-clients-and-amazon-elasticache-for-redis/) in many best practices documents for Redis clients, so I propose making this the default.

The default strategy I implemented will do the following:

* Delay will be equal to (n^2) * 50 ms where n is the current retry count, with a maximum value of 2000 ms.
* Jitter is a random value between 0 and 200 ms.

The millisecond values I choose are somewhat arbitrary, and I'm happy to change them if we think they're not right.

### Checklist

- [ ] Does `npm test` pass with this change (including linting)?
- [ ] Is the new or changed code fully tested?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

leibale commented:
Hi @qsymmachus, thanks for contributing!
[We are working on V5](https://github.com/redis/node-redis/tree/v5) which will add support for sentinel, RESP3, and client-side caching. If you can redo this change but for v5 it would be awesome! :)

qsymmachus commented:
@leibale no problem. I changed the base branch to v5 and reworked my code slightly in [4a36c6f](https://github.com/redis/node-redis/pull/2736/commits/4a36c6fea9a5ca4acf236179d89374aa6edc67ce)

leibale commented on packages/client/lib/client/socket.ts:
```diff
@@ -82,6 +82,15 @@ export default class RedisSocket extends EventEmitter {
   }
 
   #createReconnectStrategy(options?: RedisSocketOptions): ReconnectStrategyFunction {
+    const defaultStrategy = (retries: number) => {
```
I think this should be moved to a `#defaultReconnectStrategy` function and used from lines 109 and 114

qsymmachus replied:
Done in e7304ada8 👍

leibale reviewed (changes requested):
Overall looks good, just a few small comments and questions. You can ignore the nits and I'll do them myself.

qsymmachus commented on docs/client-configuration.md:
```diff
@@ -13,7 +13,7 @@
 | socket.keepAliveInitialDelay | `5000`                                   | If set to a positive number, it sets the initial delay before the first keepalive probe is sent on an idle socket                                                                                                                                   |
 | socket.tls                   |                                          | See explanation and examples [below](#TLS)                                                                                                                                                                                                          |
-| socket.reconnectStrategy     | `retries => Math.min(retries * 50, 500)` | A function containing the [Reconnect Strategy](#reconnect-strategy) logic                                                                                                                                                                           |
+| socket.reconnectStrategy     | `((retries^2) * 50 ms) + 0-200 ms`       | A function containing the [Reconnect Strategy](#reconnect-strategy) logic                                                                                                                                                                           |
```
I went with a plain english explanation of the strategy in [38eb603](https://github.com/redis/node-redis/pull/2736/commits/38eb60306b6974c5950432abd0ee542d5980748e), rather than try to figure out how to pseudo-code this in a legible way.

qsymmachus commented on docs/client-configuration.md:
```diff
@@ -35,12 +35,19 @@ When the socket closes unexpectedly (without calling `.quit()`/`.disconnect()`),
-    reconnectStrategy: retries => Math.min(retries * 50, 1000)
+    reconnectStrategy: retries => {
+        // Generate a random jitter between 0 – 200 ms:
+        const jitter = Math.floor(Math.random() * 200);
```
To me that sounds like optimizing how clients can define their own _custom_ strategy. But I see value in providing boilerplate to allow people to quickly define custom backoff + jitter strategies. Could we scope this PR to just the default strategy update, and then introduce this idea in a separate one?

leibale replied:
I guess we can do that in the future if we'll want to.. good call :)

qsymmachus commented:
@leibale I responded to your latest feedback; sorry for the delay!

qsymmachus commented:
@leibale bumping this one! Thanks for taking another look.

leibale reviewed (commented):
The code LGTM, but I'm not too sure about the documentation.. @guyroyse can you help us with it please?

qsymmachus commented:
@guyroyse @leibale can we merge this guy, then?

Thank you for your help!

leibale commented:
@qsymmachus done! thanks for taking the time to contribute, and sorry for the delays..
